Enable customizing the save statistics time interval

burnerelu opened the pull request:
This change extends the Advanced section of the Preferences menu with a new field, allowing changing the time statistics save interval. A null value will prevent recurrent saving.

This aims to provide the feature requested in issue #21285 

Tested updating the value from the web UI and reading it from the local client and the other way around. Did **not test** the actual file update time, as I did not touch the mechanism below the `saveStatistics` function.

I did not extend the lang/translation files as I wasn't sure I was supposed to. This is my first contribution to qBittorent, so I'm not familiar with the WoW

![image](https://github.com/user-attachments/assets/547fd995-ba65-4432-9c35-844c966162fa)

![image](https://github.com/user-attachments/assets/ebb51aa5-225d-4070-a5d0-c7188e32f762)

closes #21285

thalieht commented:
I would put this setting near "Save resume data interval"

thalieht commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -112,7 +112,7 @@ using namespace BitTorrent;
 const Path CATEGORIES_FILE_NAME {u"categories.json"_s};
 const int MAX_PROCESSING_RESUMEDATA_COUNT = 50;
-const int STATISTICS_SAVE_INTERVAL = std::chrono::milliseconds(15min).count();
+const int DEFAULT_STATISTICS_SAVE_INTERVAL_MIN = 15;
```
Not needed anymore, inline the default in the initialization list.

thalieht commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -5981,8 +5992,14 @@ void SessionImpl::handleSessionStatsAlert(const lt::session_stats_alert *alert)
+    if (m_saveStatisticsInterval > 0)
+    {
+        const auto saveInterval = std::chrono::milliseconds(std::chrono::minutes(m_saveStatisticsInterval));
+        if(m_statisticsLastUpdateTimer.hasExpired(saveInterval.count()))
```
```suggestion
        if (m_statisticsLastUpdateTimer.hasExpired(saveInterval.count()))
```

thalieht commented:
>Did **not test** the actual file update time

Tested, works as intended.

burnerelu commented:
Moved it according to @thalieht 's  suggestion[1].

Also moved the corresponding code around to keep the ordering of the items & fixed the rest of the remarks.

[1]:
![image](https://github.com/user-attachments/assets/9a5680b9-d6af-4593-93d6-023c2e6d532e)
![image](https://github.com/user-attachments/assets/0ae3ffec-7ee7-421b-b26d-bc04edbe113e)

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -257,6 +257,8 @@ namespace BitTorrent
         virtual void setPerformanceWarningEnabled(bool enable) = 0;
         virtual int saveResumeDataInterval() const = 0;
         virtual void setSaveResumeDataInterval(int value) = 0;
+        virtual int saveStatisticsInterval() const = 0;
+        virtual void setSaveStatisticsInterval(int value) = 0;
```
```suggestion
        virtual std::chrono::minutes saveStatisticsInterval() const = 0;
        virtual void setSaveStatisticsInterval(std::chrono::minutes value) = 0;
```
Take a look at Preferences::getWebUIBanDuration, Preferences::setWebUIBanDuration for an example.

burnerelu replied:
IMO it's better to use int, as UI elements and class storage would also be int.  std::chrono::minutes is only needed when checking the timer, reverse conversion not even needed in this case. Might be wrong since I'm not familiar with the codebase.

Will update the pull req with the changes in the following hours

burnerelu replied:
I agree that it is generally better to use `std::chrono` over int when referring to time interval, but it still feels a bit _forced_ here.  The GUI converts the `int` value of the spinbox to `std::chrono::minutes` before calling `session->setSaveStatisticsInterval(...)`, which in turn casts it back to `int` to store it. 

I'll update the pull request in a few minutes and you can take a look.

Chocobo1 commented on src/base/bittorrent/sessionimpl.h:
```diff
@@ -689,6 +691,7 @@ namespace BitTorrent
         CachedSettingValue<bool> m_isBandwidthSchedulerEnabled;
         CachedSettingValue<bool> m_isPerformanceWarningEnabled;
         CachedSettingValue<int> m_saveResumeDataInterval;
+        CachedSettingValue<int> m_saveStatisticsInterval;
```
This can be kept as is.

Chocobo1 commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -5981,8 +5991,14 @@ void SessionImpl::handleSessionStatsAlert(const lt::session_stats_alert *alert)
-        saveStatistics();
+    if (m_saveStatisticsInterval > 0)
+    {
+        const auto saveInterval = std::chrono::milliseconds(std::chrono::minutes(m_saveStatisticsInterval));
+        if (m_statisticsLastUpdateTimer.hasExpired(saveInterval.count()))
+        {
+            saveStatistics();
+        }
```
```suggestion
        const auto saveInterval = std::chrono::duration_cast<std::chrono::milliseconds>(m_saveStatisticsInterval);
        if (m_statisticsLastUpdateTimer.hasExpired(saveInterval.count()))
            saveStatistics();
```

Chocobo1 reviewed (commented):
Overall LGTM, only one small comment.

burnerelu commented on src/gui/advancedsettings.h:
```diff
@@ -74,7 +74,8 @@ private slots:
              m_spinBoxMaxConcurrentHTTPAnnounces, m_spinBoxStopTrackerTimeout, m_spinBoxSessionShutdownTimeout,
-             m_spinBoxSavePathHistoryLength, m_spinBoxPeerTurnover, m_spinBoxPeerTurnoverCutoff, m_spinBoxPeerTurnoverInterval, m_spinBoxRequestQueueSize;
+             m_spinBoxSavePathHistoryLength, m_spinBoxPeerTurnover, m_spinBoxPeerTurnoverCutoff, m_spinBoxPeerTurnoverInterval, m_spinBoxRequestQueueSize,
+             m_spinBoxSaveStatisticsInterval;
```
done

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -3518,6 +3518,16 @@ void SessionImpl::setSaveResumeDataInterval(const int value)
+    return std::chrono::minutes(m_saveStatisticsInterval);
+}
+
+void SessionImpl::setSaveStatisticsInterval(const std::chrono::minutes timeInMinutes)
```
Am I the only one for whom it seems superfluous to mention twice that the value represents minutes?

burnerelu replied:
In a previous commit, the variable type was `int`. I've omitted to change its name after going for `std::chrono::minutes`

burnerelu commented:
@Chocobo1  @glassez  is there any formal process required to merge this pull req? anything else required from my side?

Chocobo1 commented:
@burnerelu 
Thank you!

>@Chocobo1 @glassez is there any formal process required to merge this pull req? anything else required from my side?

Nah, just wait for someone's free time to merge it.
